### PR TITLE
Fix ZMQ Notification initialization and shutdown

### PR DESCRIPTION
**Pull request by promag**

### CODE REVIEW

**Comment by laanwj** [comment]
Can you explain what is the issue that this is fixing, and how it fixes it?

File: src/zmq/zmqnotificationinterface.cpp [hunk 1]
```diff
@@ -99,7 +104,7 @@ bool CZMQNotificationInterface::Initialize()
         return false;
     }
 
-    return false;
+    return true;
```
Comment (promag) [inline src/zmq/zmqnotificationinterface.cpp]: @laanwj This was wrong.

File: src/init.cpp [hunk 2]
```diff
@@ -1176,7 +1175,6 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     pzmqNotificationInterface = CZMQNotificationInterface::CreateWithArguments(mapArgs);
 
     if (pzmqNotificationInterface) {
-        pzmqNotificationInterface->Initialize();
```
Comment (promag) [inline src/init.cpp]: @laanwj The return value was discarded.

File: src/zmq/zmqnotificationinterface.cpp [hunk 3]
```diff
@@ -59,6 +58,12 @@ CZMQNotificationInterface* CZMQNotificationInterface::CreateWithArguments(const
     {
         notificationInterface = new CZMQNotificationInterface();
         notificationInterface->notifiers = notifiers;
+
+        if (!notificationInterface->Initialize())
```
Comment (promag) [inline src/zmq/zmqnotificationinterface.cpp]: @laanwj Initialisation is implicit and properly handled.

File: src/init.cpp (hunk 2 above)
Comment (promag) [inline src/init.cpp]: @laanwj Usage in `init.cpp` is simpler.

File: src/zmq/zmqnotificationinterface.cpp [hunk 4]
```diff
@@ -21,8 +21,7 @@ CZMQNotificationInterface::CZMQNotificationInterface() : pcontext(NULL)
 
 CZMQNotificationInterface::~CZMQNotificationInterface()
 {
-    // ensure Shutdown if Initialize is called
-    assert(!pcontext);
+    Shutdown();
```
Comment (promag) [inline src/zmq/zmqnotificationinterface.cpp]: @laanwj Shutdown is implicit.

File: src/zmq/zmqnotificationinterface.h [hunk 5]
```diff
@@ -19,10 +19,11 @@ class CZMQNotificationInterface : public CValidationInterface
 
     static CZMQNotificationInterface* CreateWithArguments(const std::map<std::string, std::string> &args);
 
+protected:
     bool Initialize();
```
Comment (promag) [inline src/zmq/zmqnotificationinterface.h]: @laanwj No need to expose these since they are called implicitly.

**Comment by laanwj** [comment]
Agree re: Initialize/Shutdown. As Initialize is always immediately called after the constructor and Shutdown before the destructor it makes little sense to require these to be called explicitly (and have it possible for the object to be in 'created but uninitialized' state.

**Comment by promag** [comment]
@laanwj :+1:

**Comment by jonasschnelli** [comment]
Codereview ACK.
Makes sense to use the destructor instead `init.cpp` pollution. This also fixes a possible tiny-shutdown-only memory leak in case of something went wrong during initialization.

**Comment by laanwj** [comment]
utACK.

Please do include more information in your commit message. With an empty description, someone browsing git in the future will not be able to understand the reasoning behind your change.

**Comment by promag** [comment]
@laanwj :ok_hand:

**Pull request merged**